### PR TITLE
Upgrade to SQLAlchemy 1.4 and convert endpoints to asyncio

### PR DESCRIPTION
**Pull request by jianyuan**
- [x] Upgrade to SQLAlchemy 1.4, which supports AsyncIO. [Full guide here](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#asynchronous-i-o-asyncio).
- [x] Convert API endpoints to asyncio.
- [ ] Fix Celery.

Addresses: #315

### CODE REVIEW

File: {{cookiecutter.project_slug}}/backend/app/app/tests/utils/user.py [hunk 1]
```diff
@@ -10,41 +10,43 @@
 from app.tests.utils.utils import random_email, random_lower_string
 
 
-def user_authentication_headers(
+async def user_authentication_headers(
     *, client: TestClient, email: str, password: str
```
Comment (Andrew-Sheridan) [inline {{cookiecutter.project_slug}}/backend/app/app/tests/utils/user.py:14]: Should update all these `TestClient` annotations to `AsyncClient`

File: {{cookiecutter.project_slug}}/backend/app/app/api/deps.py [hunk 2]
```diff
@@ -4,28 +4,25 @@
...
 )
 
 
-def get_db() -> Generator:
-    try:
-        db = SessionLocal()
-        yield db
-    finally:
-        db.close()
+async def get_db() -> Generator:
```
Comment (jokull) [inline {{cookiecutter.project_slug}}/backend/app/app/api/deps.py:19]: `AsyncGenerator`?

File: {{cookiecutter.project_slug}}/backend/app/pyproject.toml [hunk 3]
```diff
@@ -21,9 +21,11 @@ gunicorn = "^20.0.4"
 jinja2 = "^2.11.2"
 psycopg2-binary = "^2.8.5"
 alembic = "^1.4.2"
-sqlalchemy = "^1.3.16"
 pytest = "^5.4.1"
 python-jose = {extras = ["cryptography"], version = "^3.1.0"}
+asyncpg = "^0.22.0"
+httpx = "^0.17.1"
+SQLAlchemy = "^1.4.5"
```
Comment (jokull) [inline {{cookiecutter.project_slug}}/backend/app/pyproject.toml:28]: 1.4.10 or just ^1.4?

**Comment by Andrew-Sheridan** [comment]
@jianyuan This branch is an excellent reference!

File: {{cookiecutter.project_slug}}/backend/app/app/tests/utils/user.py (hunk 1 above)
Comment (jianyuan) [reply {{cookiecutter.project_slug}}/backend/app/app/tests/utils/user.py]: Updated, thanks!

**Comment by gmos** [comment]
Missed a couple in the start-up.
[async-startups-diff.txt](https://github.com/tiangolo/full-stack-fastapi-postgresql/files/6489514/async-startups-diff.txt)

**Comment by jokull** [comment]
Celery doesn't yet have a way to queue tasks without blocking the asyncio loop or resorting to `asgiref.sync_to_async`. Maybe it should be skipped for this tutorial? [arq](https://arq-docs.helpmanual.io) would be a better choice.

**Comment by AntonOfTheWoods** [comment]
As a warning for others, the `sqlalchemy-stubs` dev dependency is also not compatible with recent `sqlalchemy` - you will get errors in editors (`vscode` at the very least), though it will still run fine.

**Comment by jokull** [comment]
> As a warning for others, the `sqlalchemy-stubs` dev dependency is also not compatible with recent `sqlalchemy` - you will get errors in editors (`vscode` at the very least), though it will still run fine.

I think it's compatible with SQLAlchemy, but Pyright doesn't. If you want VS Code to show SQLAlchemy typing errors you're better off using the mypy extension. I configured it and it works nicely, but I disabled default typing errors.

File: {{cookiecutter.project_slug}}/backend/app/app/api/deps.py (hunk 2 above)
Comment (saffist3r) [reply {{cookiecutter.project_slug}}/backend/app/app/api/deps.py:19]: @jokull  Was added here  https://github.com/tiangolo/full-stack-fastapi-postgresql/pull/359/files#diff-8125f68fd870242defd7f66196d415f38af1d26dcb1bfbb2609556a7991ceaf1L29

**Comment by tiangolo** [comment]
Thanks @jianyuan! :bow: 

I wanted to migrate to SQLModel, it was done in another PR, that is based on SQLAlchemy 2.0. :tada: 

So, now I'll close this one, but thanks for all the effort! :cake:

**Comment by AntonOfTheWoods** [comment]
@tiangolo , if you were even remotely grateful for the effort you would have signalled something before 3 YEARS after the PR. I know, you are busy, but there were plenty of people submitting PRs and presumably many who could have spent a little effort to keep things updated. But no, a single (big name) maintainer who just ghosts everyone and their PRs for YEARS, then comes back as if it were last week...

**Pull request closed without merging**